src/flint/types/nmod_poly.pyx

```diff
@@ -621,16 +632,49 @@ cdef class nmod_poly(flint_poly):
             (3, [(x + 4, 1), (x + 2, 1), (x^2 + 4*x + 1, 1)])
 
         """
+        if algorithm is None:
+            algorithm = 'irreducible'
+        elif algorithm not in ('berlekamp', 'cantor-zassenhaus'):
+            raise ValueError(f"unknown factorization algorithm: {algorithm}")
+        return self._factor(algorithm)
+
+    def factor_squarefree(self):
+        """
+        Factors *self* into square-free polynomials. Returns (*c*, *factors*)
+        where *c* is the leading coefficient and *factors* is a list of
+        (*poly*, *exp*).
+
+            >>> x = nmod_poly([0, 1], 7)
+            >>> p = x**2 * (x/2 - 1)**2 * (x + 1)**3
+            >>> p
+            2*x^7 + 5*x^6 + 4*x^5 + 2*x^4 + 2*x^3 + x^2
+            >>> p.factor_squarefree()
+            (2, [(x^2 + 5*x, 2), (x + 1, 3)])
+            >>> p.factor()
+            (2, [(x, 2), (x + 5, 2), (x + 1, 3)])
+
+        """
+        return self._factor('squarefree')
+
+    def _factor(self, factor_type):
         cdef nmod_poly_factor_t fac
         cdef mp_limb_t lead
         cdef int i
+
         nmod_poly_factor_init(fac)
-        if algorithm == 'berlekamp':
+
+        if factor_type == 'berlekamp':
             lead = nmod_poly_factor_with_berlekamp(fac, self.val)
-        elif algorithm == 'cantor-zassenhaus':
+        elif factor_type == 'cantor-zassenhaus':
             lead = nmod_poly_factor_with_cantor_zassenhaus(fac, self.val)
-        else:
+        elif factor_type == 'irreducible':
             lead = nmod_poly_factor(fac, self.val)
+        elif factor_type == 'squarefree':
+            nmod_poly_factor_squarefree(fac, self.val)
+            lead = (<nmod>self.leading_coefficient()).val
```

Comment:
I'm not 100% sure about this.

The doc says:

> ```
> void nmod_poly_factor_squarefree(nmod_poly_factor_t res, const nmod_poly_t f)
> ```
> Sets res to a square-free factorization of f.

Whereas `nmod_poly_factor` returns the leading coefficient `nmod_poly_factor_squarefree` does not mention anything about it.

The factorisation seems to be just all monic factors though so I presume we just need to take the leading coefficient like this:
```python
In [2]: x = nmod_poly([0, 1], 7)

In [3]: (3*(x + 1)*(x + 4)**2).factor_squarefree()
Out[3]: (3, [(x + 1, 1), (x + 4, 2)])
```